video/image_writer.c

```diff
@@ -354,11 +371,34 @@ bool write_image(struct mp_image *image, const struct image_writer_opts *opts,
         destfmt = IMGFMT_RGB24;
     }
 #endif
+    if (opts->format == AV_CODEC_ID_WEBP && !opts->webp_lossless) {
+        // For lossy images, libwebp has its own RGB->YUV conversion.
+        // We don't want that, so force YUV/YUVA here.
+        int alpha = image->fmt.flags & MP_IMGFLAG_ALPHA;
+        destfmt = alpha ? pixfmt2imgfmt(AV_PIX_FMT_YUVA420P) : IMGFMT_420P;
+    }
 
     if (!destfmt)
         destfmt = get_target_format(&ctx);
 
-    struct mp_image *dst = convert_image(image, destfmt, log);
+    struct mp_image_params p = {0}, *destp = NULL;
+    if (opts->format == AV_CODEC_ID_MJPEG &&
+        (mp_imgfmt_get_desc(destfmt).flags & MP_IMGFLAG_YUV)) {
+        // Currently, assume what FFmpeg's jpg encoder needs.
+        // Of course this works only for non-HDR (no HDR support in libswscale).
+        p.color.levels = MP_CSP_LEVELS_PC;
+        p.color.space = MP_CSP_BT_601;
+        p.chroma_location = MP_CHROMA_CENTER;
+        destp = &p;
+    } else if (opts->format == AV_CODEC_ID_WEBP &&
+               (mp_imgfmt_get_desc(destfmt).flags & MP_IMGFLAG_YUV)) {
+        p.color.levels = MP_CSP_LEVELS_TV;
+        p.color.space = MP_CSP_BT_601;
+        p.chroma_location = MP_CHROMA_CENTER;
+        destp = &p;
+    }
```

Comment:
I wouldn't know where. This part is certainly needed as the levels end up wrong otherwise.

Reply:
Yeah, I was thinking about how it tries to signal the full range formats to the decoder (ffmpeg's unfortunate deprecated legacy API which you can't not use), and the hardcoded case which you actually just moved here.

I don't like how this checks for MP_IMGFLAG_YUV (something which I maybe wanted to remove, because it's so vague). Also, you have to admit that the handling of the new mp_image_params parameter is not very elegant (it seems to ignore most members, and destfmt is redundant with mp_image_params.imgfmt). With all these exceptions and hardcoded information, all that code to determine the pixel format automatically also becomes questionable.